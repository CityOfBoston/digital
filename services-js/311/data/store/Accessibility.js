// @flow

import { observable, autorun } from 'mobx';

export default class Accessibility {
  el: ?HTMLElement = null;

  @observable message: string = '';
  @observable interrupt: boolean = false;

  messageListenerDisposer: ?Function = null;

  attach() {
    this.el = document.getElementById('ariaLive');

    this.messageListenerDisposer = autorun(
      () => {
        if (this.el) {
          this.el.innerText = this.message;
          this.el.setAttribute(
            'aria-live',
            this.interrupt ? 'assertive' : 'polite'
          );
        }
      },
      {
        name: 'a11y message listener',
      }
    );
  }

  detatch() {
    this.el = null;

    if (this.messageListenerDisposer) {
      this.messageListenerDisposer();
    }
  }
}
