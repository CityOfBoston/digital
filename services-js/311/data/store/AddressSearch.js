// @flow

import { observable, action, computed, runInAction } from 'mobx';

import type { LoopbackGraphql } from '../dao/loopback-graphql';
import searchAddress from '../dao/search-address';
import reverseGeocode from '../dao/reverse-geocode';
import type { AddressUnit, SearchAddressPlace } from '../types';

// This class co-ordinates location queries between the LocationPopUp pane
// and the LocationMap map. LocationPopUp automatically updates the
// RequestForm (which contains the data that will go into the request) based
// on changes to this instance.
export default class AddressSearch {
  @observable query: string = '';
  @observable searching: boolean = false;
  @observable lastQuery: string = '';
  @observable.ref lastSearchError: ?Error = null;

  @observable.shallow places: ?Array<SearchAddressPlace> = null;
  @observable _currentPlaceIndex: number = -1;
  @observable highlightedPlaceIndex: number = -1;
  @observable _currentUnitIndex: number = 0;
  @observable
  currentReverseGeocodeLocation: ?{ lat: number, lng: number } = null;
  @observable currentReverseGeocodeLocationIsValid: boolean = true;

  @observable mode: 'search' | 'geocode' = 'search';

  // We keep track of whether the constituent seems to be pointing at a location
  // on the map or choosing from the location menu, since we have to choose to
  // send either lat/lng *or* address string to Open311. lat/lng matches the map
  // exactly, but only address string captures the unit.
  @observable intent: 'ADDRESS' | 'LATLNG' = 'ADDRESS';

  // Used so that LocationMap doesn't zoom things under the search box.
  searchPopupWidth: number = 0;

  loopbackGraphql: ?LoopbackGraphql;

  start(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;
  }

  stop() {
    this.loopbackGraphql = null;
  }

  @computed
  get currentPlaceIndex(): number {
    return this._currentPlaceIndex;
  }

  set currentPlaceIndex(currentPlaceIndex: number) {
    this._currentPlaceIndex = currentPlaceIndex;
    this.intent = 'ADDRESS';
  }

  @computed
  get currentUnitIndex(): number {
    return this._currentUnitIndex;
  }

  set currentUnitIndex(currentUnitIndex: number) {
    this._currentUnitIndex = currentUnitIndex;
    this.intent = 'ADDRESS';
  }

  @action
  async search(selectFirst: boolean): Promise<void> {
    if (!this.loopbackGraphql) {
      return;
    }

    this.places = null;
    this.lastQuery = this.query;
    this.lastSearchError = null;
    this.currentReverseGeocodeLocation = null;
    this.mode = 'search';

    if (!this.query) {
      return;
    }

    try {
      this.searching = true;
      const places = await searchAddress(this.loopbackGraphql, this.query);

      runInAction('search - searchAddress success', () => {
        this.searching = false;
        this.places = places;
        this.currentPlaceIndex = places.length === 1 || selectFirst ? 0 : -1;
        this.highlightedPlaceIndex = this.currentPlaceIndex;
        this.currentUnitIndex = 0;
        this.intent = 'ADDRESS';
      });
    } catch (err) {
      runInAction('search - searchAddress error', () => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat &&
          !err.server &&
          window._opbeat('captureException', err);
      });
    }
  }

  @computed
  get location(): ?{ lat: number, lng: number } {
    return (
      this.currentReverseGeocodeLocation ||
      (this.currentPlace ? this.currentPlace.location : null)
    );
  }

  set location(location: {| lat: number, lng: number |}) {
    this.query = '';

    if (
      this.location &&
      this.location.lat === location.lat &&
      this.location.lng === location.lng
    ) {
      return;
    }

    if (!this.loopbackGraphql) {
      return;
    }

    this.currentReverseGeocodeLocation = location;
    this.places = null;
    this.mode = 'geocode';
    this.lastQuery = '';
    this.lastSearchError = null;

    this.searching = true;
    reverseGeocode(this.loopbackGraphql, location).then(
      action('reverseGeocode success', place => {
        this.searching = false;

        if (place) {
          // We overwrite the location in the place to stay with the one picked so
          // that the map marker doesnâ€™t move.
          this.places = [
            {
              location,
              address: place.address,
              addressId: place.addressId,
              units: place.units,
            },
          ];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = true;
          this.intent = 'LATLNG';
        } else {
          this.places = [];
          this.currentPlaceIndex = 0;
          this.highlightedPlaceIndex = 0;
          this.currentUnitIndex = 0;
          this.currentReverseGeocodeLocationIsValid = false;
        }
      }),
      action('reverseGeocode error', err => {
        this.searching = false;
        this.lastSearchError = err;
        window._opbeat && window._opbeat('captureException', err);
      })
    );
  }

  @computed
  get currentPlace(): ?SearchAddressPlace {
    return (
      (this.places &&
        this.places.length > this.currentPlaceIndex &&
        this.places[this.currentPlaceIndex]) ||
      null
    );
  }

  @computed
  get currentUnit(): ?AddressUnit {
    return (
      (this.currentPlace &&
        this.currentPlace.units.length > this.currentUnitIndex &&
        this.currentPlace.units[this.currentUnitIndex]) ||
      null
    );
  }

  @computed
  get notFound(): boolean {
    return !!this.places && this.places.length === 0;
  }

  @computed
  get address(): string {
    if (this.currentUnit) {
      return this.currentUnit.address;
    } else if (this.currentPlace) {
      return this.currentPlace.address;
    } else {
      return '';
    }
  }

  @computed
  get addressId(): ?string {
    if (this.currentUnit) {
      return this.currentUnit.addressId;
    } else if (this.currentPlace) {
      return this.currentPlace.addressId;
    } else {
      return null;
    }
  }

  @computed
  get units(): Array<AddressUnit> {
    return this.currentPlace ? this.currentPlace.units : [];
  }
}
