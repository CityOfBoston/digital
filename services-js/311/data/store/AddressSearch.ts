import { observable, action, computed, runInAction } from 'mobx';

import { FetchGraphql, GraphqlError } from '@cityofboston/next-client-common';

import searchAddress from '../queries/search-address';
import reverseGeocode from '../queries/reverse-geocode';
import {
  AddressUnit,
  SearchAddressPlace,
  ReverseGeocodeResult,
} from '../types';

// This class co-ordinates location queries between the LocationPopUp pane
// and the LocationMap map. LocationPopUp automatically updates the
// RequestForm (which contains the data that will go into the request) based
// on changes to this instance.
export default class AddressSearch {
  @observable query: string = '';
  @observable searching: boolean = false;
  @observable lastQuery: string = '';
  @observable.ref lastSearchError: Error | null = null;

  @observable.shallow private _places: Array<SearchAddressPlace> | null = null;
  @observable private _currentPlaceIndex: number = -1;
  @observable highlightedPlaceIndex: number = -1;
  @observable private _currentUnitIndex: number = 0;
  @observable
  currentReverseGeocodeLocation: { lat: number; lng: number } | null = null;
  @observable currentReverseGeocodeLocationIsValid: boolean = true;

  @observable mode: 'search' | 'geocode' = 'search';

  // We keep track of whether the constituent seems to be pointing at a location
  // on the map or choosing from the location menu, since we have to choose to
  // send either lat/lng *or* address string to Open311. lat/lng matches the map
  // exactly, but only address string captures the unit.
  @observable intent: 'ADDRESS' | 'LATLNG' = 'ADDRESS';

  // Used so that LocationMap doesn't zoom things under the search box.
  searchPopupWidth: number = 0;

  private fetchGraphql: FetchGraphql | null = null;

  start(fetchGraphql: FetchGraphql) {
    this.fetchGraphql = fetchGraphql;
  }

  stop() {
    this.fetchGraphql = null;
  }

  @computed
  get places(): Array<SearchAddressPlace> | null {
    return this._places;
  }

  clearPlaces() {
    this._places = null;
  }

  setPlaces(
    places: Array<SearchAddressPlace> | null,
    mode: 'search' | 'geocode',
    selectFirst: boolean
  ) {
    this._places = places;

    // These calls are not going through the setters to avoid changing the intent.
    this._currentPlaceIndex =
      places && (places.length === 1 || selectFirst) ? 0 : -1;
    this.highlightedPlaceIndex = this._currentPlaceIndex;
    this._currentUnitIndex = 0;

    if (
      mode === 'search' &&
      (!this.currentPlace || !this.currentPlace.alwaysUseLatLng)
    ) {
      this.intent = 'ADDRESS';
    } else {
      this.intent = 'LATLNG';
    }
  }

  @computed
  get currentPlaceIndex(): number {
    return this._currentPlaceIndex;
  }

  set currentPlaceIndex(currentPlaceIndex: number) {
    if (currentPlaceIndex === this._currentPlaceIndex) {
      return;
    }

    this._currentPlaceIndex = currentPlaceIndex;
    // We specifically just set the private variable because the
    // currentUnitIndex setter forces the intent to "ADDRESS," which is not what
    // we want if we're picking an address or something else with
    // alwaysUseLatLng.
    this._currentUnitIndex = 0;

    // Your intent should likely already be 'ADDRESS' in this case, because
    // picking from multiple places means that you did a forward search and the
    // intent was already address. Regardless, since you chose something by
    // name, we update the intent unless it’s a place that should never be
    // referenced by name (which for now means an intersection).
    const place = this._places && this._places[currentPlaceIndex];
    if (place) {
      this.intent = place.alwaysUseLatLng ? 'LATLNG' : 'ADDRESS';
    }
  }

  @computed
  get currentUnitIndex(): number {
    return this._currentUnitIndex;
  }

  set currentUnitIndex(currentUnitIndex: number) {
    this._currentUnitIndex = currentUnitIndex;
    // If you're changing the unit, we need to send the address string along so
    // the backend can seach for which unit you chose. This won't be called for
    // intersections because they don't have units.
    this.intent = 'ADDRESS';
  }

  @action
  async search(selectFirst: boolean): Promise<void> {
    const { fetchGraphql } = this;

    if (!fetchGraphql) {
      return;
    }

    this.clearPlaces();
    this.lastQuery = this.query;
    this.lastSearchError = null;
    this.currentReverseGeocodeLocation = null;
    this.mode = 'search';

    if (!this.query) {
      return;
    }

    try {
      this.searching = true;
      const places = await searchAddress(fetchGraphql, this.query);

      (runInAction as any)('search - searchAddress success', () => {
        this.searching = false;
        this.setPlaces(places, 'search', selectFirst);
      });
    } catch (err) {
      (runInAction as any)('search - searchAddress error', () => {
        this.searching = false;
        this.lastSearchError = err;

        window.Rollbar && err.source !== 'server' && window.Rollbar.error(err);
      });
    }
  }

  @computed
  get location(): { lat: number; lng: number } | null {
    return (
      this.currentReverseGeocodeLocation ||
      (this.currentPlace ? this.currentPlace.location : null)
    );
  }

  set location(location: { lat: number; lng: number } | null) {
    if (!location) {
      this.clearPlaces();
      return;
    }

    const { fetchGraphql } = this;

    this.query = '';

    if (
      this.location &&
      this.location.lat === location.lat &&
      this.location.lng === location.lng
    ) {
      return;
    }

    if (!fetchGraphql) {
      return;
    }

    this.currentReverseGeocodeLocation = location;
    this.clearPlaces();
    this.mode = 'geocode';
    this.lastQuery = '';
    this.lastSearchError = null;

    this.searching = true;
    reverseGeocode(fetchGraphql, location).then(
      action('reverseGeocode success', (place: ReverseGeocodeResult) => {
        this.searching = false;

        if (place) {
          // We overwrite the location in the place to stay with the one picked so
          // that the map marker doesn’t move.
          this.setPlaces(
            [
              {
                location,
                address: place.address,
                addressId: place.addressId,
                exact: place.exact,
                units: place.units,
                alwaysUseLatLng: place.alwaysUseLatLng,
              },
            ],
            'geocode',
            true
          );

          this.currentReverseGeocodeLocationIsValid = true;
        } else {
          this.setPlaces([], 'geocode', true);
          this.currentReverseGeocodeLocationIsValid = false;
        }
      }),
      action('reverseGeocode error', (err: GraphqlError) => {
        this.searching = false;
        this.lastSearchError = err;
        window.Rollbar && err.source !== 'server' && window.Rollbar.error(err);
      })
    );
  }

  @computed
  get currentPlace(): SearchAddressPlace | null {
    return (
      (this.places &&
        this.places.length > this.currentPlaceIndex &&
        this.places[this.currentPlaceIndex]) ||
      null
    );
  }

  @computed
  get currentUnit(): AddressUnit | null {
    return (
      (this.currentPlace &&
        this.currentPlace.units.length > this.currentUnitIndex &&
        this.currentPlace.units[this.currentUnitIndex]) ||
      null
    );
  }

  @computed
  get notFound(): boolean {
    return !!this.places && this.places.length === 0;
  }

  @computed
  get address(): string {
    if (this.currentUnit) {
      return this.currentUnit.address;
    } else if (this.currentPlace) {
      return this.currentPlace.address;
    } else {
      return '';
    }
  }

  @computed
  get addressId(): string | null {
    if (this.currentUnit) {
      return this.currentUnit.addressId;
    } else if (this.currentPlace) {
      return this.currentPlace.addressId;
    } else {
      return null;
    }
  }

  @computed
  get units(): Array<AddressUnit> {
    return this.currentPlace ? this.currentPlace.units : [];
  }
}
