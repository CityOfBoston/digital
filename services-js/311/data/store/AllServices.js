// @flow

import { observable } from 'mobx';

const GROUPS = [
  {
    name: 'Animals',
    id: 'Animals',
    description: 'Pet Ownership, Problems with Animals, Rodents & Pests',
  },
  {
    name: 'Fees and Pricing',
    id: 'Fees-Pricing',
    description: 'NEEDS DESCRIPTION',
  },
  {
    name: 'Health, Safety, and Social Services',
    id: 'Health-Safety',
    description:
      'Crime, Fire Prevention & Control, Healthy Living & Inspections',
  },
  {
    name: 'Parks and Public Space',
    id: 'Parks-Public',
    description: 'Park Maintenance, Permits, Tree Planting, Donations',
  },
  {
    name: 'Property, Housing and Construction',
    id: 'Property-Construction',
    description:
      'Home Ownership, Renting, Inspections, Code Enforcement, Permitting, Complaints',
  },
  {
    name: 'Schools',
    id: 'Schools',
    description: 'NEEDS DESCRIPTION',
  },
  {
    name: 'Snow or Weather Related Concerns',
    id: 'Snow',
    description: 'Plowing, snow removal',
  },
  {
    name: 'Streets, Sidewalks, Traffic, and Vehicles',
    id: 'Streets-Sidewalks',
    description:
      'Signs & Signals, Parking, Driving, Public Transportation, Taxi Issues, Road Maintenance',
  },
  {
    name: 'Trash, Recycling, and Graffiti',
    id: 'Trash-Graffiti',
    description:
      'Trash, Recycling, Graffiti, Dumping, Street Cleaning, Neighborhood Cleanup',
  },
  {
    name: 'Utilities',
    id: 'Utilities',
    description: 'NEEDS DESCRIPTION',
  },
  {
    name: 'Other Services',
    id: 'Other',
    description: 'General service requests that donâ€™t fit in another category',
  },
];

export class Group {
  name: string;
  id: string;
  description: string;

  @observable open: boolean = false;

  constructor({
    name,
    id,
    description,
  }: {
    name: string,
    id: string,
    description: string,
  }) {
    this.name = name;
    this.id = id;
    this.description = description;
  }
}

/* This is kept in the store so that clicking "back" to the page keeps the
   categories open. */
export default class AllServices {
  groups: Group[] = GROUPS.map(g => new Group(g));
  otherGroup: Group = new Group({
    name: 'Uncategorized Services',
    id: '',
    description: 'These services have not yet been categorized',
  });
}
