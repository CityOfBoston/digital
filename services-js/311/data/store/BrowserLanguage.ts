import { IncomingMessage } from 'http';
import languageParser from 'accept-language-parser';

export type LanguagePreference = {
  code: string;
  region: string | null;
  quality: number;
};

/**
 * Parses the "Accept-Language" header from the given request and returns the
 * array of LanguagePreference objects.
 */
export default function parseLanguagePreferences(
  req: IncomingMessage | undefined
): Array<LanguagePreference> {
  // process.browser check to allow dead code removal to clear out the parsing
  // module in the client build, where webpack has that set to true.
  return req && !process.browser
    ? languageParser.parse(req.headers['accept-language'])
    : [];
}
