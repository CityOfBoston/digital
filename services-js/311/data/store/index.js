// @flow

import { observable, computed, useStrict } from 'mobx';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { Service, ServiceSummary, SubmittedRequest } from '../types';
import Question from './Question';

// MobX will enforce that state changes only happen in action blocks.
useStrict(true);

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{| lat: number, lng: number |} = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

export class AppStore {
  @observable description: string = '';

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @observable.shallow serviceSummaries: ServiceSummary[];
  serviceCache: Map<string, Service> = observable.shallowMap({});

  apiKeys: {[service: string]: string} = {};
  isPhone: boolean = false;

  @observable requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  @observable.ref _currentService: ?Service = null;
  @observable.ref currentServiceError: ?Object = null;

  @computed get currentService(): ?Service {
    return this._currentService;
  }

  set currentService(service: ?Service) {
    if (service === this._currentService) {
      return;
    }

    this._currentService = service;
    this.currentServiceError = null;
    this.questions = [];

    if (service) {
      try {
        this.contactInfo.required = service.contactRequired;
        this.locationInfo.required = service.locationRequired;
        this.questions = Question.buildQuestions(service.attributes);
      } catch (e) {
        this.currentServiceError = e;
      }
    }
  }

  @computed
  get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
