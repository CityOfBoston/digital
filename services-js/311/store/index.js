// @flow
/* global window */
/* eslint no-underscore-dangle: 0 */

import { createStore, compose, applyMiddleware } from 'redux';
import type { Store as ReduxStore } from 'redux';
import { middleware as reduxPackMiddleware } from 'redux-pack';
import thunk from 'redux-thunk';
import inject from 'redux-inject';

import type { RequestAdditions } from '../server/next-handlers';
import { reducer } from './modules';
import type { AppState, AppAction } from './modules';
import makeLoopbackGraphql from './deps/loopback-graphql';
import type { LoopbackGraphql } from './deps/loopback-graphql';

// Re-export these to keep things centralized in the “store” module.
export type State = AppState;
export type Action = AppAction;

export type Store = ReduxStore<State, Action>;

export type Deps = {|
  loopbackGraphql: LoopbackGraphql,
|}

// We implement our own Dispatch type to account for the actions and return
// values provided by our middleware. (The Redux built-in type is from Action
// (which must have a type property) to Action. We need to take Actions that
// don't have properties and return pretty much anything, since that’s what
// thunks can return.
export type Dispatch = (action: Action) => any;

// For the browser, we keep a singleton store that can be re-used across pages.
// A module global is the only mechanism for sharing between Next.js pages.
let browserStore = null;

// Factored out to allow for injecting test dependencies and state while
// maintaining the same middleware.
export function makeStore(deps: ?Deps = null, initialState: ?State = null): Store {
  // Support for Redux Devtools Extension: https://github.com/zalmoxisus/redux-devtools-extension
  const composeEnhancers = (process.browser && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

  const enhancers = composeEnhancers(
    applyMiddleware(
      inject(deps),
      thunk,
      reduxPackMiddleware,
    ),
  );

  if (initialState) {
    return createStore(reducer, initialState, enhancers);
  } else {
    return createStore(reducer, enhancers);
  }
}

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 *
 * req is the Node request (expected to contain injected dependencies for server
 * data fetching) that should be provided when creating the store for the
 * server’s getInitialProps calls. Can be null when server rendering or on the
 * client.
 */
export default function getStore(req: ?RequestAdditions, initialState: ?State = null): Store {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const deps = {
    loopbackGraphql: makeLoopbackGraphql(req),
  };

  const store = makeStore(deps, initialState);

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
