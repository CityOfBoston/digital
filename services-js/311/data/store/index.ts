/* global window */

import { observable, configure as mobxConfigure, action } from 'mobx';
import Router from 'next/router';

import RequestSearch from './RequestSearch';
import AddressSearch from './AddressSearch';
import Ui from './Ui';
import BrowserLocation from './BrowserLocation';
import AllServices from './AllServices';
import {
  FetchGraphql,
  ScreenReaderSupport,
  GaSiteAnalytics,
  RouterListener,
} from '@cityofboston/next-client-common';
import { SiteAnalytics } from '@cityofboston/next-client-common/build/SiteAnalytics';

// MobX will enforce that state changes only happen in action blocks.
mobxConfigure({ enforceActions: true });

export type LanguagePreference = {
  code: string;
  region: string | null;
  quality: number;
};

export class AppStore {
  screenReaderSupport: ScreenReaderSupport = new ScreenReaderSupport();
  requestSearch: RequestSearch = new RequestSearch();
  ui: Ui = new Ui();
  browserLocation: BrowserLocation = new BrowserLocation();
  addressSearch: AddressSearch = new AddressSearch();
  allServices: AllServices = new AllServices();
  siteAnalytics: SiteAnalytics = new GaSiteAnalytics();

  languages: LanguagePreference[] = [];

  @observable
  liveAgentAvailable: boolean = (typeof window !== 'undefined' &&
    window.LIVE_AGENT_AVAILABLE) ||
  false;

  _liveAgentButtonId: string = '';

  get liveAgentButtonId(): string {
    return this._liveAgentButtonId;
  }

  set liveAgentButtonId(liveAgentButtonId: string) {
    this._liveAgentButtonId = liveAgentButtonId;

    if (typeof window !== 'undefined') {
      if (window.liveagent) {
        this.listenForLiveAgentEvents(liveAgentButtonId);
      } else {
        if (!window._laq) {
          window._laq = [];
        }

        window._laq.push(() =>
          this.listenForLiveAgentEvents(liveAgentButtonId)
        );
      }
    }
  }

  listenForLiveAgentEvents(liveAgentButtonId: string) {
    window.liveagent!.addButtonEventHandler(
      liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(
  fetchGraphql: FetchGraphql | null = null
): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
    store.screenReaderSupport.attach();
    store.ui.attach();
    store.browserLocation.attach(fetchGraphql);

    new RouterListener().attach({
      router: Router,
      siteAnalytics: store.siteAnalytics,
      screenReaderSupport: store.screenReaderSupport,
    });
  }

  return store;
}
