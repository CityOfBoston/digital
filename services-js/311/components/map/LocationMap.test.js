// @flow

import * as React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { runInAction } from 'mobx';

import { AppStore } from '../../data/store';

import type LocationMapClass from './LocationMap';

let LocationMap: Class<LocationMapClass>;

let previousBrowserValue;

beforeAll(() => {
  previousBrowserValue = process.browser;
  process.browser = true;

  // We need these shenanigans to get LocationMap loaded with process.browser
  // true, so that it requires leaflet, which can't be required server-side.
  LocationMap = require.requireActual('./LocationMap').default;
});

afterAll(() => {
  process.browser = previousBrowserValue;
});

test('rendering active', () => {
  const store = new AppStore();
  store.apiKeys.mapbox = {
    accessToken: 'FAKE_MAPBOX_ACCESS_TOKEN',
    styleUrl: 'mapbox://styles/fake-style',
  };

  const component = renderer.create(
    <LocationMap
      L={require('mapbox.js')}
      mapboxgl={null}
      store={store}
      mode="picker"
      mobile={false}
    />,
    { createNodeMock: () => document.createElement('div') }
  );
  expect(component.toJSON()).toMatchSnapshot();
});

describe('mounted map', () => {
  let store;
  let wrapper;
  let locationMap: LocationMap;

  beforeEach(() => {
    store = new AppStore();
    store.apiKeys.mapbox = {
      accessToken: 'FAKE_MAPBOX_ACCESS_TOKEN',
      styleUrl: 'mapbox://styles/fake-style',
    };

    wrapper = mount(
      <LocationMap
        L={require('mapbox.js')}
        mapboxgl={null}
        store={store}
        mode="picker"
        mobile={false}
      />
    );

    locationMap = wrapper.instance();
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('positions a marker based on the form location', () => {
    runInAction(() => {
      store.addressSearch.places = [
        {
          address: '1 City Hall Plaza',
          addressId: '12345',
          units: [],
          location: {
            lat: 42,
            lng: -71,
          },
        },
        {
          address: '2 City Hall Plaza',
          addressId: '12346',
          units: [],
          location: {
            lat: 43,
            lng: -70,
          },
        },
      ];
      store.addressSearch.currentPlaceIndex = 0;
    });

    const { locationSearchMarkers } = locationMap;
    const locationSearchMarker = locationSearchMarkers[0];

    if (!locationSearchMarker) {
      expect(locationSearchMarker).toBeDefined();
      return;
    }

    const markerLatLng = locationSearchMarker.getLatLng();
    expect(markerLatLng.lat).toEqual(42);
    expect(markerLatLng.lng).toEqual(-71);
  });
});
