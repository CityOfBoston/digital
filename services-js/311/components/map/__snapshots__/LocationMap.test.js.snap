// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering active 1`] = `
<div
  className={
    Object {
      "data-css-wi5peg": "",
    }
  }
  style={
    Object {
      "opacity": 1,
    }
  }
/>
`;
